generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -164,6 +166,157 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
         depth, fuzzer)));
   }
 
+  /**
+   * Function to generate an opaque zero or one by bitwise shifting left or right by an
+   * arbitrary number of bits. Generates an opaque value if the type of zero to be generated is an
+   * integer and the shading language in use supports bitwise operations.
+   *
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from bitwise shifting.
+   */
+  private Optional<Expr> opaqueZeroOrOneBitwiseShift(BasicType type, boolean constContext,
+                                                     final int depth, Fuzzer fuzzer,
+                                                     boolean isZero) {
+    if (!BasicType.allIntegerTypes().contains(type)) {
+      return Optional.empty();
+    }
+    if (!shadingLanguageVersion.supportedBitwiseOperations()) {
+      return Optional.empty();
+    }
+    final BinOp operator = generator.nextBoolean() ? BinOp.SHL : BinOp.SHR;
+    // We need to make sure the LHS and RHS of the shift expression are based on the same type
+    // explicitly, so we can't generate opaque values to shift by as they will implicitly be
+    // considered signed integers, and both sides of the shift must be the same type.
+    final Expr shiftValue;
+    if (isZero) {
+      // We can't shift zero reliably if the number is signed because of sign bit extension, so
+      // we simply shift it zero if it's signed.
+      shiftValue = type.getElementType() == BasicType.INT
+          ? new IntConstantExpr("0")
+          : new UIntConstantExpr(generator.nextInt(1000) + "u");
+    } else {
+      assert !isZero;
+      // We can't shift one at all without changing it.
+      shiftValue = type.getElementType() == BasicType.INT
+          ? new IntConstantExpr("0")
+          : new UIntConstantExpr("0u");
+    }
+    return Optional.of(
+        new ParenExpr(
+            new BinaryExpr(
+                makeOpaqueZeroOrOne(isZero, type, constContext, depth, fuzzer),
+                shiftValue, operator)));
+  }
+
+  /**
+   * Function to generate an opaque value by performing a bitwise operation on an opaque zero or
+   * an opaque one.
+   * Possibilities to create zero include:
+   * Bitwise ANDing an opaque zero or one with an opaque zero: (opaque zero or one) & (opaque zero)
```

Comment:
^ Why not bitwise AND a fuzzed expression with an opaque zero?  (More general.)  And I presume you're doing it in both directions.

Reply:
Done.